### PR TITLE
Handle LegacyChoiceListAdapter being passed to ModelsToArrayTransformer (Fixes #3107)

### LINKED ISSUE #3107: Form type sonata_type_model break with Symfony 2.7.2
Using master version of Sonata-Admin with Symfony version 2.7.2 results in error:
```
Argument 1 passed to Sonata\AdminBundle\Form\DataTransformer\ModelsToArrayTransformer::__construct() must be an instance of Sonata\AdminBundle\Form\ChoiceList\ModelChoiceList, instance of Symfony\Component\Form\ChoiceList\LegacyChoiceListAdapter given
```
when using sonata_type_model in form as of BC break in Symfony2 2.7.2:
https://github.com/symfony/symfony/commit/6325b4caf6eeeb56cbecffb1af417c2c7f14c1fb

Wrapping variable assigment $this->choiceList in ModelsToArrayTransformer in something like that should do the work:
```php
public function __construct($choiceList)
    {
        if ($choiceList instanceof LegacyChoiceListAdapter){
            $this->choiceList = $choiceList->getAdaptedList();
        }else if ($choiceList instanceof ModelChoiceList){
            $this->choiceList = $choiceList;
        }
        else{
            throw new \InvalidArgumentException('Argument 1 must be an instance of Sonata\AdminBundle\Form\ChoiceList\ModelChoiceList or Symfony\Component\Form\ChoiceList\LegacyChoiceListAdapter');
        }
    }
```

### PR DESCRIPTION
**Pull request by aderuwe**

### CODE REVIEW

**Comment by aderuwe** [comment]
The travis failure is unrelated, it seems.

**Comment by npotier** [comment]
Hi @aderuwe thanks for this PR :)

**Comment by jsampedro77** [comment]
Yes, thank you. Hope this will be merged soon.

**Comment by tunne** [comment]
Thx @aderuwe

**Comment by rzymek01** [comment]
I also had this problem and I resolved it by changing one line of code in file "admin-bundle/Form/ChoiceList/ModelChoiceList.php"
before:
class ModelChoiceList extends SimpleChoiceList
after:
class ModelChoiceList extends ArrayChoiceList

(add use statement: use Symfony\Component\Form\ChoiceList\ArrayChoiceList;)

**Comment by OskarStark** [comment]
@rande @Soullivaneuh can you restart the travis build please?

File: Form/DataTransformer/ModelsToArrayTransformer.php [hunk 1]
```diff
@@ -23,14 +23,21 @@
  */
 class ModelsToArrayTransformer implements DataTransformerInterface
 {
+    /**
+     * @var object
```
Comment (OskarStark) [inline Form/DataTransformer/ModelsToArrayTransformer.php]: this comment is not necessary

File: Form/DataTransformer/ModelsToArrayTransformer.php [hunk 2]
```diff
@@ -23,14 +23,21 @@
...
+     */
     protected $choiceList;
 
     /**
-     * @param \Sonata\AdminBundle\Form\ChoiceList\ModelChoiceList $choiceList
+     * @param object $choiceList
      */
-    public function __construct(ModelChoiceList $choiceList)
+    public function __construct($choiceList)
```
Comment (OskarStark) [inline Form/DataTransformer/ModelsToArrayTransformer.php]: isn't there an interface which is used by both of the ```choiceLists```?

completely removing the type hint is not so good IMO

**Comment by rande** [comment]
@OskarStark done

**Comment by aderuwe** [comment]
I'll look at those remarks tomorrow. Regarding the interface, I did try it, but can't remember what it failed on in that case...

**Comment by aderuwe** [comment]
I tried @rzymek01's suggestion as well, but equally can't remember what that failed on. :) I'll give both a shot again tomorrow.

**Comment by aderuwe** [comment]
@rande saw on twitter today that more people/projects are experiencing issues with travis/composer...

File: Form/DataTransformer/ModelsToArrayTransformer.php (hunk 2 above)
Comment (Armstrong1992) [inline Form/DataTransformer/ModelsToArrayTransformer.php]: That's why I proposed:
```
if ($choiceList instanceof LegacyChoiceListAdapter){
            $this->choiceList = $choiceList->getAdaptedList();
        }else if ($choiceList instanceof ModelChoiceList){
            $this->choiceList = $choiceList;
        }
        else{
            throw new \InvalidArgumentException('Argument 1 must be an instance of Sonata\AdminBundle\Form\ChoiceList\ModelChoiceList or Symfony\Component\Form\ChoiceList\LegacyChoiceListAdapter');
        }
```

File: Form/DataTransformer/ModelsToArrayTransformer.php (hunk 1 above)
Comment (Soullivaneuh) [inline Form/DataTransformer/ModelsToArrayTransformer.php]: It should. But class should be specified instead of `object`.

Comment (Soullivaneuh) [inline Form/DataTransformer/ModelsToArrayTransformer.php]: Should be `ModelChoiceList|LegacyChoiceListAdapter`.

File: Form/DataTransformer/ModelsToArrayTransformer.php [hunk 3]
```diff
@@ -23,14 +23,21 @@
...
 class ModelsToArrayTransformer implements DataTransformerInterface
 {
+    /**
+     * @var object
+     */
     protected $choiceList;
 
     /**
-     * @param \Sonata\AdminBundle\Form\ChoiceList\ModelChoiceList $choiceList
+     * @param object $choiceList
```
Comment (Soullivaneuh) [inline Form/DataTransformer/ModelsToArrayTransformer.php]: Should be `ModelChoiceList|LegacyChoiceListAdapter`.

**Comment by Soullivaneuh** [comment]
ping @aderuwe can you answer the review? Thanks.

**Comment by aderuwe** [comment]
@Soullivaneuh brought the typehint back, no idea why I thought it wouldn't work. It does. That fixes the docblock as well.

As for making `ModelChoiceList` extend `ArrayChoiceList`, that didn't work because then the id's become the values in the eventual form.

**Comment by Soullivaneuh** [comment]
@aderuwe please rebase your PR to get Travis working.

See #3117 for more details about the issue.

**Comment by aderuwe** [comment]
@Soullivaneuh done!

**Comment by aderuwe** [comment]
Ah, those failures do relate to this PR now - I'll take a look.

**Comment by aderuwe** [comment]
I can't spend more time on it right now, but those failures are weird... I don't know where we are to get the `getIdentifier() ` etc from now. Perhaps I just miss something, will take a look at home.

**Comment by npotier** [comment]
Thanks again @aderuwe. I hope your work will be merged soon.

**Comment by Soullivaneuh** [comment]
Your Travis issue is here:

```php
$this->modelChoiceList = $this->getMockBuilder('Symfony\Component\Form\ChoiceList\ChoiceListInterface');
```

`getIdentifier` is proper to sonata `ModelChoiceList`, not symfony choice list.

I'm not sure your PR is a good approach. I think we should fix the deprecated class extension instead.

I already created a PR a time ago about it: https://github.com/sonata-project/SonataAdminBundle/pull/3022

BTW, I think we should search or created an issue on Symfony because of BC introduction.

I think it's related to this PR but not sure: https://github.com/symfony/symfony/pull/15061, @webmozart could you please confirm?

ping @rande

**Comment by aderuwe** [comment]
@Soullivaneuh you're right, the proper way is probably to make Sonata's `SimpleChoiceList` extend `ArrayChoiceList` and fix the behaviour issues... (id's becoming values in the form)

I don't have very much time to look into that right now though ...

**Comment by Soullivaneuh** [comment]
@aderuwe I'll take a look if #3022 will provide a solution. Waiting for Symfony answers too.

**Comment by Soullivaneuh** [comment]
Ok it's separated issues. Your PR is good I was on mistake.

I'll take a look at your tests.

**Comment by Soullivaneuh** [comment]
But I think we should mark conflict for Symfony `2.7.0` and `2.7.1` because of lot of BC breaking on those versions...

**Comment by Soullivaneuh** [comment]
I think we should not mock up `LegacyAdapter` but instanciate it with the `ModelChoiceList` mock as argument.

**Comment by Soullivaneuh** [comment]
Got it! I'll make a PR on yours! ;-)

**Comment by aderuwe** [comment]
Thank you @Soullivaneuh

**Comment by Soullivaneuh** [comment]
I have some issue with older symfony version now... working on it.

**Comment by OskarStark** [comment]
Thanks for your work @Soullivaneuh

**Comment by Soullivaneuh** [comment]
Done: https://github.com/aderuwe/SonataAdminBundle/pull/1

By chance, this seems to work also for `2.7.0` and `2.7.1`.

**Comment by aderuwe** [comment]
@Soullivaneuh merged and rebased - seems like somehow it lost your authorship there though...

File: Form/DataTransformer/ModelsToArrayTransformer.php [hunk 4]
```diff
@@ -12,6 +12,8 @@
 namespace Sonata\AdminBundle\Form\DataTransformer;
 
 use Sonata\AdminBundle\Form\ChoiceList\ModelChoiceList;
+use Symfony\Component\Form\ChoiceList\ChoiceListInterface;
```
Comment (Soullivaneuh) [inline Form/DataTransformer/ModelsToArrayTransformer.php]: This import is not used anymore. My missing.

CS will cry, remove it.

**Comment by Soullivaneuh** [comment]
> seems like somehow it lost your authorship there though...

It's normal, only one author per commit. With the squash, first commit has priority.

Not a problem for me, I will continue to sleep well. :+1:

File: Form/DataTransformer/ModelsToArrayTransformer.php (hunk 4 above)
Comment (aderuwe) [inline Form/DataTransformer/ModelsToArrayTransformer.php]: Done

**Comment by Soullivaneuh** [comment]
\o/

Thank you @aderuwe.

**Comment by aderuwe** [comment]
@Soullivaneuh thank you, too!

**Comment by samsser** [comment]
When will the  Sandbox-Sonata-Build  will be updated pls ?

**Comment by Soullivaneuh** [comment]
> When will the Sandbox-Sonata-Build will be updated pls ?

ping @rande

**Pull request merged**